### ISSUE TITLE
backward of torch.repeat slower than for torch.repeat_interleave

### ISSUE THREAD

**Issue by tobiasfshr** [pytorch/pytorch#43192]
## 🐛 Bug

I noticed a high performance difference when repeating a variable that requires grad. If I repeat the variable with torch.repeat, the backward pass will take 10x more time than when using torch.repeat_interleave. Below you find a minimal example for reproduction, however here i found only a ~2x difference.

## To Reproduce
Code Snippet:
https://gist.github.com/tobiasfshr/c4e3473caf9a758312a16a6b5301f456

repeat_interleave:
iter time:  0.0019862818717956542

repeat:
iter time:  0.004346790313720703

## Expected behavior

As far as i understand the doc, both methods should behave the same w.r.t. gradient computation, although they have slightly different interfaces.

## Environment
 - PyTorch Version (e.g., 1.0): 1.6.0
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): -
 - Python version: 3.6
 - CUDA/cuDNN version: 10.1 / 7.6.5
 - GPU models and configuration: Problem appears on both GTX 1660 (CUDA 10.1) and Tesla V100 (CUDA 10.0)
 - Any other relevant information:

cc @ezyang @SsnL @albanD @zou3519 @gqchen @VitalyFedyunin @ngimel

**Comment by zou3519** [comment]
I can reproduce this, but I'm not sure if it is worth fixing. The repeat pattern being used here is repeating a Tensor along some dimension of size 1. The best thing to actually do here is to expand the tensors along a dimension to avoid a copy; replacing the `repeat` in the benchmark code with a `expand` produces the best performance on my machine:

```
z = torch.rand((1, 32)).requires_grad_()

repeated = z.repeat(1024, 1)
repeated = z.repeat_interleave(1024, dim=0)
repeated = z.expand(1024, 32)
```
@tobiasfshr: are you repeating a tensor along some dimension that does not have size 1? `repeat` and `repeat_interleave` have different semnatics in that case.

**Comment by ngimel** [comment]
Regardless of a particular case, backward of `repeat` is inefficient due to https://github.com/pytorch/pytorch/blob/e7564b076c13325fba1704d9d07844a26041f10f/torch/csrc/autograd/FunctionsManual.cpp#L645. `sum_tensorlist` is inefficiently implemented as a sequential sum. `grad` is guaranteed to be evenly split into necessary-size pieces, so instead of chunk + sum_tensorlist `grad` tensor can be reshaped and `sum` can be called on it, that should give similar performance to`expand`.

**Comment by tobiasfshr** [comment]
@zou3519 Thanks for the suggestion! I'm actually repeating the tensor in my code along a dimension with size > 1. Specifically, I have a tensor x with size B*N, C and the tensor y that is to be repeated, which has size B, C. To concatenate them, i repeat the tensor N / B times. However, you can solve this also in this way by expanding y:
`y.unsqueeze(1).expand(B, N, C).contiguous().view(-1, C)`
